### PR TITLE
Full node changes required for CAT / Standalone

### PR DESCRIPTION
**Pull request by Yostra**

### CODE REVIEW

File: chia/consensus/blockchain.py
Comment (arvidn) [inline chia/consensus/blockchain.py:301]: the next version of `clvm_rs` will actually always include the 3rd field here, but it will be an empty bytes buffer for coins that didn't specify it. So you'll need to check `condition.vars[2] != b""` as well

File: chia/consensus/blockchain.py
Comment (arvidn) [inline chia/consensus/blockchain.py:173]: where is the call to this function where you do something with the additional return values?

File: chia/consensus/blockchain.py
Comment (Yostra) [reply chia/consensus/blockchain.py]: changed

File: chia/consensus/blockchain.py
Comment (Yostra) [reply chia/consensus/blockchain.py]: `receive_block_batch` and `respond_block` in ful_node.py call this function and then pass the results to `update_wallets`

File: chia/consensus/blockchain.py
Comment (mariano54) [inline chia/consensus/blockchain.py:304]: @Quexington was saying that the 3rd argument should be a list, so we can add more things there later. This is also how I am setting the memo for retirement of CAT tokens: 
https://github.com/Chia-Network/chia-blockchain/pull/8600

File: chia/consensus/blockchain.py
Comment (mariano54) [inline chia/consensus/blockchain.py:434]: Can you add a type annotation here? Not sure what type this list is.

File: chia/consensus/blockchain.py
Comment (mariano54) [inline chia/consensus/blockchain.py:297]: coin_id is type bytes32 instead of bytes

File: chia/consensus/blockchain.py
Comment (mariano54) [inline chia/consensus/blockchain.py:437]: coin_id instead of coin_name

File: chia/consensus/blockchain.py
Comment (mariano54) [inline chia/consensus/blockchain.py:438]: Can we remove the "key" variable?

File: chia/full_node/hint_store.py
Comment (mariano54) [inline chia/full_node/hint_store.py:34]: coin_id is bytes32.

File: chia/consensus/blockchain.py
Comment (mariano54) [inline chia/consensus/blockchain.py:307]: We should only add the memos which are less than 32 bytes right? Otherwise this is a DOS vector.
I suggest the following naming: The list of bytes in the 3rd argument should be called the memos, and the memos are stored as hints only if they are <= 32 bytes.

Comment (Yostra) [reply chia/consensus/blockchain.py:307]: clvm_rs doesn't return anything longer than 32bytes

File: chia/consensus/blockchain.py
Comment (Yostra) [reply chia/consensus/blockchain.py:438]: avoids hashing same thing multiple times

File: chia/consensus/blockchain.py
Comment (Yostra) [reply chia/consensus/blockchain.py:304]: I think this needs to be changed in clvm_rs first, was there change there?

Comment (Yostra) [reply chia/consensus/blockchain.py:304]: Just saw Arvids pr is open, I'll have to wait for that to get in & clvm_rs pypy release is available.

File: chia/full_node/full_node_api.py
Comment (arvidn) [inline chia/full_node/full_node_api.py:1412]: we can receive this request from an untrusted peer, right? as someone not very familiar with `get_ses_heights()`, it's hard to say whether this is safe, or whether an untrusted peer can make us do an unreasonable amount of work.

`get_ses_heights()` in particular, looks scary, as it doesn't take any parameters.

File: chia/full_node/hint_store.py
Comment (arvidn) [inline chia/full_node/hint_store.py:20]: if we ever want to support multiple hints per coin, changing this schema may be the biggest hurdle

File: chia/full_node/hint_store.py
Comment (arvidn) [inline chia/full_node/hint_store.py:25]: this function is called `get_hints()` but returns coin-IDs. It seems better names would be: `lookup_hints()` or `get_coin_ids_from_hints()`

File: chia/protocols/protocol_message_types.py
Comment (arvidn) [inline chia/protocols/protocol_message_types.py:98]: what's `ses` short for?

File: chia/protocols/shared_protocol.py
Comment (arvidn) [inline chia/protocols/shared_protocol.py:8]: do we record which messages are available in what version of the protocol anywhere?

File: tests/wallet/simple_sync/test_simple_sync_protocol.py
Comment (arvidn) [inline tests/wallet/simple_sync/test_simple_sync_protocol.py:529]: it seems like you should check the actual value of the `coin_states` too, not just that you got *something*

**Review by arvidn — COMMENTED** [review]
perhaps the tuple of coin updates and new hints should be put in a new class holding state updates/subscription updates. Such a class could also do the deduplication and aggregation that you're currently doing "ad hoc" (or, sprinkled into the existing code)

File: chia/consensus/blockchain.py
Comment (Yostra) [reply chia/consensus/blockchain.py:401]: yea, it just saved on indentation bellow

File: chia/full_node/hint_store.py
Comment (Yostra) [reply chia/full_node/hint_store.py:20]: good catch, I'll add another auto-increment id column there so that coin id is not unique

File: chia/protocols/protocol_message_types.py
Comment (Yostra) [reply chia/protocols/protocol_message_types.py:98]: sub epoch summary

File: tests/wallet/simple_sync/test_simple_sync_protocol.py
Comment (Yostra) [reply tests/wallet/simple_sync/test_simple_sync_protocol.py:529]: added

File: chia/protocols/shared_protocol.py
Comment (Yostra) [reply chia/protocols/shared_protocol.py:8]: No yet

File: chia/full_node/hint_store.py
Comment (Yostra) [reply chia/full_node/hint_store.py]: added

File: chia/consensus/blockchain.py
Comment (Yostra) [reply chia/consensus/blockchain.py:175]: I took a shortcut here, receive_block get's called on 100s of places in tests and if another arg was added in the call I would have to refactor all of those.

Comment (arvidn) [reply chia/consensus/blockchain.py:175]: so maybe a dedicated type for updates would make sense then. then new updates could be added without affecting this signature

**Comment by arvidn** [comment]
I retriggered the failing test. Is there a risk this patch slowed down the execution to cause that timing constraint to be exceeded?
If so, we could bump it further under the understanding that we're increasing the tech debt

File: chia/consensus/blockchain.py
Comment (mariano54) [reply chia/consensus/blockchain.py:175]: I'm not sure how this is passing the tests, given that the response is checked in a lot of tests.
I agree with arvid, this response is getting quite complex. Perhaps we can put all of these things into one struct (including the fork_height). 

We can leave this for the future, but this is a pretty important method, so it would be good to keep it simple

**Comment by arvidn** [comment]
Reviewing this, I haven't found any obvious errors, but I'm also not really convinced that all new edge cases introduced by this are covered by tests.

File: chia/consensus/blockchain.py
Comment (mariano54) [reply chia/consensus/blockchain.py:403]: We don't run it again. If the npc_result already exists (it's Optional, and not present when syncing) then we will not run the generator. So we are guaranteed to only run it once.

File: chia/full_node/full_node_api.py
Comment (mariano54) [reply chia/full_node/full_node_api.py:1412]: Blockchain.get_ses_heights() is very fast since it stores these things in memory (the height of each sub-epoch)

Comment (mariano54) [reply chia/full_node/full_node_api.py:1412]: Everything here seems safe and very fast. However, we should have a rate limit like other messages.

**Review by mariano54 — APPROVED** [review]
LGTM after addressing arvid's feedback (and maybe adding a rate limit)

**Pull request merged**
